File: 2Arrays and Strings/Code/Python/2-2Combine.py
def combine(arr1, arr2):
    # ans is the answer
    ans = []
    i = j = 0
    while i < len(arr1) and j < len(arr2):
        if arr1[i] < arr2[j]:
            ans.append(arr1[i])
            i += 1
        else:
            ans.append(arr2[j])
            j += 1
    
    while i < len(arr1):
        ans.append(arr1[i])
        i += 1
    
    while j < len(arr2):
        ans.append(arr2[j])
        j += 1
    
    return ans

# test-cases
print(combine([1,4,7,20], [3,5,6]))  # [1,3,4,5,6,7,20]

print(combine([4,8,20], [1,5,6,17])) # [1,4,5,6,8,17,20]

print(combine([5,6,8,9], [1, 2, 4])) # [1,2,4,5,6,8,9]

# Time-complexity: O(n+m) where
## n: length of array 1
## m: length of array 2

# Space-complexity: O(1)

File: 2Arrays and Strings/Code/Cpp/2-2Combine.cpp
#include<iostream>
#include<vector>

using namespace std;

vector<int> combine(vector<int>& arr1, vector<int>& arr2) {
    // ans is the answer
    vector<int> ans;
    int i = 0, j = 0;
    while (i < arr1.size() && j < arr2.size()) {
        if (arr1[i] < arr2[j]) {
            ans.push_back(arr1[i]);
            i++;
        } else {
            ans.push_back(arr2[j]);
            j++;
        }
    }
    
    while (i < arr1.size()) {
        ans.push_back(arr1[i]);
        i++;
    }
    
    while (j < arr2.size()) {
        ans.push_back(arr2[j]);
        j++;
    }
    
    return ans;
}

int main()
{
    // test-case #1
    vector<int> s1 = {1,4,7,20};
    vector<int> s2 = {3,5,6};
    vector<int> result = combine(s1, s2);
    cout << "{";
    for (int i = 0; i < result.size(); i++) {
        cout << result[i];
        if (i != result.size() - 1) {
            cout << ",";
        }
    }
    cout << "}" << endl; // {1,3,4,5,6,7,20}

    // test-case #2
	s1 = {4, 8, 20};
	s2 = {1,5,6,17};
	result = combine(s1, s2);
	cout << "{";
	for (int i = 0; i < result.size(); i++) {
		cout << result[i];
		if (i != result.size() - 1) {
			cout << ",";
		}
	}
	cout << "}" << endl; // {1,4,5,6,8,17,20}

    // test-case #3
    s1 = {5,6,8,9};
	s2 = {1,2,4};
	result = combine(s1, s2);
	cout << "{";
	for (int i = 0; i < result.size(); i++) {
		cout << result[i];
		if (i != result.size() - 1) {
			cout << ",";
		}
	}
	cout << "}" << endl; // {1,2,4,5,6,8,9}
	
    return 0;
}

// Time-complexity : O(n+m)
// Space complexity: O(1)
// Note: We usually don't count the output as extra space.
